Compiler/src/abstractinterpretation.jl

```diff
@@ -2139,22 +2139,32 @@ function abstract_call_builtin(interp::AbstractInterpreter, f::Builtin, (; fargs
     return rt
 end
 
-function form_partially_defined_struct(@nospecialize(obj), @nospecialize(name))
+function form_partially_defined_struct(𝕃ᵢ::AbstractLattice, @nospecialize(obj), @nospecialize(name))
     obj isa Const && return nothing # nothing to refine
     name isa Const || return nothing
     objt0 = widenconst(obj)
     objt = unwrap_unionall(objt0)
     objt isa DataType || return nothing
     isabstracttype(objt) && return nothing
+    objt <: Tuple && return nothing
     fldidx = try_compute_fieldidx(objt, name.val)
     fldidx === nothing && return nothing
-    isa(obj, PartialStruct) && return define_field(obj, fldidx)
+    if isa(obj, PartialStruct)
+        obj.undefs[fldidx] === false && return nothing
+        newundefs = copy(obj.undefs)
+        newundefs[fldidx] = false
+        return PartialStruct(𝕃ᵢ, obj.typ, newundefs, copy(obj.fields))
+    end
     nminfld = datatype_min_ninitialized(objt)
-    fldidx > nminfld || return nothing
-    undef = partialstruct_init_undef(objt, fldidx; all_defined = false)
-    undef[fldidx] = false
-    fields = Any[fieldtype(objt0, i) for i = 1:fldidx]
-    return PartialStruct(fallback_lattice, objt0, undef, fields)
+    fldidx ≤ nminfld && return nothing
+    fldcnt = fieldcount_noerror(objt)::Int
+    fields = Any[fieldtype(objt0, i) for i = 1:fldcnt]
+    if fields[fldidx] === Union{}
+        return nothing # refine this field
```

Comment:
I'm not sure I understand the comment here, doesn't refining mean returning a `PartialStruct`? (though it makes sense to return `nothing` as a `Union{}` field type already implies `undef[fldidx] === true`)

Reply:
This PR wasn't finished yesterday, so the logic might have been messed up. Undef the refined definition of `PartialStruct`, `fields[fldidx] === Union{}` is a special case, meaning that field is guaranteed not to be initialized, so we need to bail out here.